include/deal.II/grid/grid_tools.h

```diff
@@ -1285,6 +1285,269 @@ namespace GridTools
+                        typename dealii::Triangulation<dim>::cell_iterator>>
+              potential_boxes_and_cells;
+
+            for (const auto &box_and_cell : boxes_and_cells)
+              if (cell_marked(box_and_cell.second))
+                potential_boxes_and_cells.emplace_back(box_and_cell);
+
+            const auto marked_cell_tree = pack_rtree(potential_boxes_and_cells);
+
+
+            // process requests
+            for (unsigned int i = 0; i < request.size(); ++i)
+              {
+                const auto &request_index          = request[i].first;
+                const auto &requested_intersection = request[i].second;
+
+                // create a bounding box with tolerance
+                const auto bb = BoundingBox<spacedim>(request[i].second)
+                                  .create_extended(tolerance);
+
+                for (const auto &[box, cell] :
+                     marked_cell_tree |
+                       boost::geometry::index::adaptors::queried(
+                         boost::geometry::index::intersects(bb)))
+                  {
+#ifdef DEAL_II_WITH_CGAL
+                    auto const &try_intersection =
+                      dealii::CGALWrappers::get_vertices_in_cgal_order(
+                        cell, cache.get_mapping());
+
+                    const auto &found_intersections = dealii::CGALWrappers::
+                      compute_intersection_of_cells<dim, request_dim, spacedim>(
+                        try_intersection, requested_intersection, tolerance);
+
+                    if (found_intersections.size() > 0)
+                      {
+                        for (const auto &found_intersection :
+                             found_intersections)
+                          {
+                            answer.emplace_back(request_index,
+                                                found_intersection);
+
+                            send_components.emplace_back(
+                              std::make_pair(cell->level(), cell->index()),
+                              other_rank,
+                              request_index,
+                              found_intersection,
+                              requested_intersection);
+                          }
+                      }
+#else
+                    Assert(false, ExcNeedsCGAL());
+#endif
+                  }
+              }
+          }
+
+        if (answer.size() > 0)
+          send_ranks.emplace_back(other_rank);
+
+        return answer;
+      };
+
+      const auto process_answer = [&](const unsigned int other_rank,
+                                      const AnswerType & answer) {
+        for (unsigned int i = 0; i < answer.size(); ++i)
+          recv_components.emplace_back(other_rank,
+                                       answer[i].first,
+                                       answer[i].second);
+
+        if (answer.size() > 0)
+          recv_ranks.emplace_back(other_rank);
```

Comment:
... and also this.

Reply:
Same discussion in https://github.com/dealii/dealii/pull/15165